### PR TITLE
[SFTTrainer]: Add Aux Loss for MoE models.

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?
The `SFTTrainer` currently doesn't account for the load-balancing/auxilliary loss commonly used in training MoE models to ensure that all experts have approximately the same number of tokens routed to them.

This PR adds this loss to the final loss if enabled by the model's config.

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -499,7 +499,9 @@ class SFTTrainer(Trainer):
               [`~transformers.PreTrainedModel.save_pretrained`], e.g., `'./my_model_directory/'`. The model is loaded
               using [`~transformers.AutoModelForCausalLM.from_pretrained`] with the keyword arguments in
               `args.model_init_kwargs`.
-            - A [`~transformers.PreTrainedModel`] object. Only causal language models are supported.
+            - A [`~transformers.PreTrainedModel`] object. Only causal language models are supported. If you're training
+            a MoE model and want to include the load balancing/auxilliary loss as a part of the final loss, remember to set
+            the `output_router_logits` config of the model to `True`.
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py]: that's not true anymore: 

-> "Only causal (vision) language models  and are supported."

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -1103,6 +1120,8 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
                 total_sum = total_tokens.sum()
                 accuracy = (correct_tokens.sum() / total_sum).item() if total_sum > 0 else 0.0
                 self._metrics[mode]["mean_token_accuracy"].append(accuracy)
+                if self.aux_loss_enabled:
+                    self._metrics[mode]["aux_loss"].append(aux_loss.item())
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py]: you need to gather the aux loss before logging

**Comment by qgallouedec** [comment]
Super cool PR! Only a few remarks, and we're good to merge end test!

File: trl/trainer/sft_trainer.py (hunk 2 above)
Comment (pramodith) [reply trl/trainer/sft_trainer.py]: Done!

**Review by qgallouedec — APPROVED** [review]
lgtm!
